extendr-engine/src/lib.rs

```diff
@@ -71,6 +71,13 @@ pub fn with_r(f: impl FnOnce()) {
     // is no `end_r()` call here.
 }
 
+#[ctor::dtor]
+fn shutdown_r() {
+    if START_R.is_completed() {
+        end_r();
+    }
+}
```

Comment:
If I understand correctly, this means `shitdown_r()` is always called regardless the user already ended the session (e.g. by calling `end_r()` by themselves) or not. Is it safe to call `end_r()` multiple times?

Reply:
I hope the is completed will ensure that end r is only called if the cell was called.
That happens only once.
And the shutdown is in a dtor which is a function that is executed only when extendr engine is shutdown.

Reply:
Yes, I know shutdown_r() is ensured to be called only once. But, as end_r() is a public function, users can call it freely, right?

Reply:
Right. I should also make it private.
It's not safe to call it multiple times.

Reply:
Or, probably like start_r(), you can wrap it with std::sync::Once().

Reply:
The thing is, `test!` is defined in `extendr-api`, which uses `start_r`.
And it is done like that because of reasons due to.. keeping out non-API symbols from the library file.
Atleast in theory;

Reply:
I don't think `end_r` is safe to run multiple times at all. I don't think `test!` should contain `end_r` either.
I think it should be private and with this, end_r is appropriately ran at the end of the program.

Unless one is doing something particularly tricky, with threads, and each thread spawning its own process.. then this setup will suffice maybe.

Reply:
Thanks, makes sense.